crates/matrix-sdk-ui/src/timeline/pagination.rs

```diff
@@ -67,49 +69,45 @@ impl super::Timeline {
-
-                Err(EventCacheError::BackpaginationError(
-                    PaginatorError::InvalidPreviousState {
-                        actual: PaginatorState::Paginating, ..
-                    },
-                )) => {
-                    warn!("Another pagination request is already happening, returning early");
-                    return Ok(false);
-                }
-
-                Err(err) => return Err(err),
-            };
-
-            let BackPaginationOutcome { events, reached_start } = event_cache_outcome;
-
-            let num_events = events.len();
-            trace!("Back-pagination succeeded with {num_events} events");
-
-            self.inner
-                .add_events_at(events, TimelineEnd::Front, RemoteEventOrigin::Pagination)
-                .await;
-
-            if reached_start {
-                return Ok(true);
+        let reached_start = pagination
+            .run_backwards(
+                batch_size,
+                |BackPaginationOutcome { events, reached_start }| async move {
+                    let num_events = events.len();
+                    trace!("Back-pagination succeeded with {num_events} events");
+
+                    // TODO(hywan): Remove, and let spread events via
+                    // `matrix_sdk::event_cache::RoomEventCacheUpdate` from
+                    // `matrix_sdk::event_cache::RoomPagination::run_backwards`.
+                    self.inner
+                        .add_events_at(events, TimelineEnd::Front, RemoteEventOrigin::Pagination)
+                        .await;
+
+                    if num_events == 0 && !reached_start {
+                        // As an exceptional contract: if there were no events in the response,
+                        // and we've not hit the start of the timeline, retry until we get
+                        // some events or reach the start of the timeline.
+                        return ControlFlow::Continue(());
+                    }
+
+                    ControlFlow::Break(reached_start)
+                },
+            )
+            .await;
+
+        match reached_start {
+            Ok(reached_start) => Ok(reached_start),
+
+            Err(EventCacheError::BackpaginationError(PaginatorError::InvalidPreviousState {
+                actual: PaginatorState::Paginating,
+                ..
+            })) => {
+                warn!("Another pagination request is already happening, returning early");
+                Ok(false)
             }
 
-            if num_events == 0 {
-                // As an exceptional contract: if there were no events in the response,
-                // and we've not hit the start of the timeline, retry until we get
-                // some events or reach the start of the timeline.
-                continue;
-            }
-
-            // Exit the inner loop, and ask for another limit.
-            break;
+            Err(err) => Err(err),
         }
```

Comment:
Looks like this match could be simplified to explicitly handle the intricate `Err(...)` case, and have a `result => result` nistead of `Ok(reached_start) => Ok(reached_start)` and `Err(err) => Err(err)`?